industrial_ci/util.sh

```diff
@@ -51,93 +51,76 @@ function ici_time_start {
     TRAVIS_END_TIME=$(date +%s%N)
     TIME_ELAPSED_SECONDS=$(( ($TRAVIS_END_TIME - $TRAVIS_START_TIME)/1000000000 ))
     echo -e "ici_time:end:$TRAVIS_TIME_ID:start=$TRAVIS_START_TIME,finish=$TRAVIS_END_TIME,duration=$(($TRAVIS_END_TIME - $TRAVIS_START_TIME))\e[0K"
     echo -e "ici_fold:end:$TRAVIS_FOLD_NAME\e[${color_wrap}m<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\e[0m"
-    echo -e "\e[0K\e[${color_wrap}mFunction $TRAVIS_FOLD_NAME took $(( $TIME_ELAPSED_SECONDS / 60 )) min $(( $TIME_ELAPSED_SECONDS % 60 )) sec\e[0m"
+    echo -e "\e[0K\e[${color_wrap}mFunction $TRAVIS_FOLD_NAME returned with code '${exit_code}' after $(( $TIME_ELAPSED_SECONDS / 60 )) min $(( $TIME_ELAPSED_SECONDS % 60 )) sec \e[0m"
 
     unset $TRAVIS_FOLD_NAME
     if [ "$DEBUG_BASH" ] && [ "$DEBUG_BASH" == true ]; then set -x; fi
 }
 
 #######################################
-# This private function can exit the shell process, as well as wrapping up the timer section on Travis CI. Internally this does:
-#
-# * wraps the section that is started by ici_time_start function.
-# * resets signal handler for ERR to the bash default one, when `exit_code` is any error code that exits the shell. This allows subsequent signal handlers for ERR if any to be unaffected by any handlers defined beforehand.
-# * exits the process if non -1 value is passed to `exit_code`.
+# exit function with handling for EXPECT_EXIT_CODE
 #
+# Globals:
+#   (None)
 # Arguments:
-#  exit_code (default: -1): Unix signal. If -1 then the process continues without exiting.
-#  color_wrap (default: 32): Color code for the section delimitter text.
+#   exit_code (default: $?)
+# Returns:
+#   (None)
 #######################################
-function _end_fold_script {
-    if [ "$DEBUG_BASH" ] && [ "$DEBUG_BASH" == true ]; then set +x; fi
-    exit_code=${1:--1}  # If 1st arg is not passed, set -1.
-    color_wrap=${2:-32}
+function ici_exit {
+    local exit_code=${1:-$?}  # If 1st arg is not passed, set last error code.
+    trap - EXIT # Reset signal handler since the shell is about to exit.
 
-    if [ $exit_code -eq "1" ]; then color_wrap=31; fi  # Red color
-    if [ -z $TRAVIS_FOLD_NAME ]; then
-        ici_time_end $color_wrap
-    else
-	echo "Previous Travis fold name not found. It might be either successful termination of the script, or wrong call. Skipping 'ici_time_end' anyway."
+    # end fold if needed
+    if [ -n "$TRAVIS_FOLD_NAME" ]; then
+        if [ $exit_code -ne "0" ]; then color_wrap=31; fi  # Red color for errors
+        ici_time_end "$color_wrap" "$exit_code"
     fi
 
-    if [ $exit_code -eq "1" ]; then trap - ERR; fi  # Reset signal handler since the shell is about to exit.
-    if [ "$DEBUG_BASH" ] && [ "$DEBUG_BASH" == true ]; then set -x; fi
-    if [ $exit_code -ne "-1" ]; then exit $exit_code; fi
+    if [ "$exit_code" == "${EXPECT_EXIT_CODE:-0}" ]; then
+        exit 0
+    elif [ "$exit_code" == "0" ]; then # 0 was not expected
```

Comment:
No, this is another use case.

If exit_code != EXPECT_EXIT_CODE  then the exit_code will be used.
However, If exit_code is 0, but EXPECT_EXIT_CODE is not 0, the exit_code can not be returned as 0 would mean success.